common/rate-limiter/src/lib.rs

```diff
@@ -0,0 +1,392 @@
+    pub fn open() -> Self {
+        Self {
+            buckets: RwLock::new(HashMap::new()),
+            default_bucket_size: std::usize::MAX,
+            default_fill_rate: std::usize::MAX,
+            open: true,
+        }
+    }
+
+    /// Retrieve bucket, or create a new one
+    pub fn bucket(&mut self, key: Key) -> Arc<Mutex<Bucket>> {
+        if !self.open {
+            self.bucket_inner(key, |size, rate| {
+                Arc::new(Mutex::new(Bucket::new(size, rate)))
+            })
+        } else {
+            self.bucket_inner(key, |_, _| Arc::new(Mutex::new(Bucket::open())))
+        }
+    }
+
+    fn bucket_inner<F: FnOnce(usize, usize) -> Arc<Mutex<Bucket>>>(
+        &mut self,
+        key: Key,
+        bucket_create: F,
+    ) -> Arc<Mutex<Bucket>> {
+        // Attempt to do a weaker read lock first, followed by a write lock if it's missing
+        // For the common (read) case, there should be higher throughput
+        let maybe_value = { self.buckets.read().unwrap().get(&key).cloned() };
+
+        if let Some(value) = maybe_value {
+            value
+        } else {
+            let size = self.default_bucket_size;
+            let rate = self.default_fill_rate;
+            // Write in a bucket, but make sure again that it isn't there first
+            self.buckets
+                .write()
+                .unwrap()
+                .entry(key)
+                .or_insert_with(|| bucket_create(size, rate))
+                .clone()
+        }
+    }
+
+    /// Fill the buckets with the given rate, saturating at the max size
+    /// This should occur on an interval basis
+    fn refill(&mut self) {
+        self.buckets
+            .write()
+            .unwrap()
+            .values_mut()
+            .for_each(|bucket| {
+                bucket.lock().unwrap().refill();
+            });
+    }
+
+    /// Garbage collects unused references (there's only one reference, the map)
+    fn garbage_collect(&mut self) {
+        let mut write = self.buckets.write().unwrap();
+
+        let keys_to_remove: Vec<_> = write
+            .keys()
+            .filter(|key| {
+                if let Some(arc) = write.get(key) {
+                    Arc::strong_count(arc) == 1
+                } else {
+                    false
+                }
+            })
+            .cloned()
+            .collect();
+
+        for key in keys_to_remove {
+            write.remove(&key);
+        }
+    }
+}
+
+/// A token bucket object that keeps track of everything related to a key
+/// This can be used as a standalone rate limiter; however, to make it more useful
+/// it should be wrapped in an `Arc` and a `Mutex` to be shared across threads.
+#[derive(Debug)]
+pub struct Bucket {
+    /// The current number of available tokens to be used
+    tokens: usize,
+    /// Maximum number of `tokens` in the bucket
+    size: usize,
+    /// The fill rate of the bucket.  Amount added to `tokens` on a `refill`
+    rate: usize,
```

Comment:
probably clarify the units on `rate`. it looks like `tokens/sec`?

Reply:
Yeah, tokens/sec